compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -0,0 +1,1282 @@
+    /// Each element of the Vec represents a scope with every scope together making
+    /// up all currently visible definitions.
+    scopes: Vec<FxHashMap<DefinitionId, Value>>,
+
+    /// True if we've expanded any macros into any functions and will need
+    /// to redo name resolution & type checking for that function.
+    changed_functions: FxHashSet<FuncId>,
+
+    /// True if we've expanded any macros into global scope and will need
+    /// to redo name resolution & type checking for everything.
+    changed_globally: bool,
+
+    in_loop: bool,
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) enum Value {
+    Unit,
+    Bool(bool),
+    Field(FieldElement),
+    I8(i8),
+    I32(i32),
+    I64(i64),
+    U8(u8),
+    U32(u32),
+    U64(u64),
+    String(Rc<String>),
+    Function(FuncId, Type),
+    Closure(HirLambda, Vec<Value>, Type),
+    Tuple(Vec<Value>),
+    Struct(FxHashMap<Rc<String>, Value>, Type),
+    Pointer(Shared<Value>),
+    Array(Vector<Value>, Type),
+    Slice(Vector<Value>, Type),
+    Code(Rc<BlockExpression>),
+}
+
+/// The possible errors that can halt the interpreter.
+#[allow(unused)]
+#[derive(Debug)]
+pub(crate) enum InterpreterError {
+    ArgumentCountMismatch { expected: usize, actual: usize, call_location: Location },
+    TypeMismatch { expected: Type, value: Value, location: Location },
+    NoValueForId { id: DefinitionId, location: Location },
+    IntegerOutOfRangeForType { value: FieldElement, typ: Type, location: Location },
+    ErrorNodeEncountered { location: Location },
+    NonFunctionCalled { value: Value, location: Location },
+    NonBoolUsedInIf { value: Value, location: Location },
+    NonBoolUsedInConstrain { value: Value, location: Location },
+    FailingConstraint { message: Option<Value>, location: Location },
+    NoMethodFound { object: Value, typ: Type, location: Location },
+    NonIntegerUsedInLoop { value: Value, location: Location },
+    NonPointerDereferenced { value: Value, location: Location },
+    NonTupleOrStructInMemberAccess { value: Value, location: Location },
+    NonArrayIndexed { value: Value, location: Location },
+    NonIntegerUsedAsIndex { value: Value, location: Location },
+    NonIntegerIntegerLiteral { typ: Type, location: Location },
+    NonIntegerArrayLength { typ: Type, location: Location },
+    NonNumericCasted { value: Value, location: Location },
+    IndexOutOfBounds { index: usize, length: usize, location: Location },
+    ExpectedStructToHaveField { value: Value, field_name: String, location: Location },
+    TypeUnsupported { typ: Type, location: Location },
+    InvalidValueForUnary { value: Value, operator: &'static str, location: Location },
+    InvalidValuesForBinary { lhs: Value, rhs: Value, operator: &'static str, location: Location },
+    CastToNonNumericType { typ: Type, location: Location },
+
+    // Perhaps this should be unreachable! due to type checking also preventing this error?
```

Comment:
We would have to add some testing but if type checking prevents it I think these are fine to unreachable!